File: src/components/schedule/cron-utils.ts
export type FrequencyType = 'hourly' | 'daily' | 'weekly' | 'monthly' | 'custom';
export type IntervalUnit = 'minutes' | 'hours';
export type MonthlyMode = 'each' | 'onThe';
export type OrdinalPosition = 'first' | 'second' | 'third' | 'fourth' | 'last';

export interface RecurringState {
  frequency: FrequencyType;
  // Hourly
  intervalValue: number;
  intervalUnit: IntervalUnit;
  // Daily
  dailyInterval: number;
  // Weekly
  weeklyInterval: number;
  weekdays: number[]; // 0=Sun, 1=Mon, ..., 6=Sat
  // Monthly
  monthlyInterval: number;
  monthlyMode: MonthlyMode;
  monthDays: number[]; // 1-31
  ordinalPosition: OrdinalPosition;
  ordinalWeekday: number; // 0=Sun, ..., 6=Sat
  // Time (used by daily, weekly, monthly)
  hour: number;
  minute: number;
  // Custom
  customExpression: string;
}

export const DEFAULT_RECURRING: RecurringState = {
  frequency: 'daily',
  intervalValue: 1,
  intervalUnit: 'hours',
  dailyInterval: 1,
  weeklyInterval: 1,
  weekdays: [1], // Monday
  monthlyInterval: 1,
  monthlyMode: 'each',
  monthDays: [1],
  ordinalPosition: 'first',
  ordinalWeekday: 1, // Monday
  hour: 9,
  minute: 0,
  customExpression: '0 * * * *',
};

const ORDINAL_MAP: Record<OrdinalPosition, string> = {
  first: '1',
  second: '2',
  third: '3',
  fourth: '4',
  last: 'L',
};

export function buildCron(state: RecurringState): string {
  switch (state.frequency) {
    case 'hourly': {
      if (state.intervalUnit === 'minutes') {
        return state.intervalValue === 1
          ? '* * * * *'
          : `*/${state.intervalValue} * * * *`;
      }
      // hours
      return state.intervalValue === 1
        ? '0 * * * *'
        : `0 */${state.intervalValue} * * *`;
    }
    case 'daily': {
      const dayPart = state.dailyInterval > 1 ? `*/${state.dailyInterval}` : '*';
      return `${state.minute} ${state.hour} ${dayPart} * *`;
    }
    case 'weekly': {
      const days = state.weekdays.length > 0 ? state.weekdays.sort((a, b) => a - b).join(',') : '1';
      return `${state.minute} ${state.hour} * * ${days}`;
    }
    case 'monthly': {
      const monthPart = state.monthlyInterval > 1 ? `*/${state.monthlyInterval}` : '*';
      if (state.monthlyMode === 'each') {
        const days = state.monthDays.length > 0 ? state.monthDays.sort((a, b) => a - b).join(',') : '1';
        return `${state.minute} ${state.hour} ${days} ${monthPart} *`;
      }
      // "On the" mode: use weekday#ordinal syntax
      const ord = ORDINAL_MAP[state.ordinalPosition];
      if (ord === 'L') {
        return `${state.minute} ${state.hour} * ${monthPart} ${state.ordinalWeekday}L`;
      }
      return `${state.minute} ${state.hour} * ${monthPart} ${state.ordinalWeekday}#${ord}`;
    }
    case 'custom':
      return state.customExpression;
  }
}

export function parseCron(expression: string): RecurringState {
  const parts = expression.trim().split(/\s+/);
  if (parts.length !== 5) {
    return { ...DEFAULT_RECURRING, frequency: 'custom', customExpression: expression };
  }

  const [minField, hourField, domField, monField, dowField] = parts;

  // Try to detect hourly: minute or hour field has */N or *, and dom/mon/dow are all *
  if (domField === '*' && monField === '*' && dowField === '*') {
    // Hourly with minutes interval: */N * * * * or * * * * *
    if (hourField === '*') {
      const minInterval = parseInterval(minField);
      if (minInterval !== null) {
        return {
          ...DEFAULT_RECURRING,
          frequency: 'hourly',
          intervalUnit: 'minutes',
          intervalValue: minInterval,
        };
      }
    }

    // Hourly with hours interval: 0 */N * * * or N N * * *
    if (minField.match(/^\d+$/) && hourField.match(/^\*\/?\d*$/)) {
      const hourInterval = parseInterval(hourField);
      if (hourInterval !== null && hourInterval > 1) {
        return {
          ...DEFAULT_RECURRING,
          frequency: 'hourly',
          intervalUnit: 'hours',
          intervalValue: hourInterval,
        };
      }
      if (hourInterval === 1) {
        return {
          ...DEFAULT_RECURRING,
          frequency: 'hourly',
          intervalUnit: 'hours',
          intervalValue: 1,
        };
      }
    }

    // Daily: M H */N * * or M H * * *
    if (minField.match(/^\d+$/) && hourField.match(/^\d+$/)) {
      const dailyInterval = parseInterval(domField);
      if (dailyInterval !== null) {
        return {
          ...DEFAULT_RECURRING,
          frequency: 'daily',
          hour: parseInt(hourField),
          minute: parseInt(minField),
          dailyInterval,
        };
      }
    }
  }

  // Weekly: M H * * 0,1,3 (dow is not * and dom is *)
  if (domField === '*' && dowField !== '*' && !dowField.includes('#') && !dowField.includes('L')) {
    const min = parseInt(minField);
    const hour = parseInt(hourField);
    if (!isNaN(min) && !isNaN(hour) && monField === '*') {
      const weekdays = dowField.split(',').map(Number).filter((n) => !isNaN(n));
      if (weekdays.length > 0) {
        return {
          ...DEFAULT_RECURRING,
          frequency: 'weekly',
          hour,
          minute: min,
          weekdays,
        };
      }
    }
  }

  // Monthly "each": M H 1,15 * * or M H 1,15 */N *
  if (domField !== '*' && !domField.includes('/') && (dowField === '*')) {
    const min = parseInt(minField);
    const hour = parseInt(hourField);
    if (!isNaN(min) && !isNaN(hour)) {
      const monthDays = domField.split(',').map(Number).filter((n) => !isNaN(n) && n >= 1 && n <= 31);
      if (monthDays.length > 0) {
        const monthlyInterval = parseInterval(monField) ?? 1;
        return {
          ...DEFAULT_RECURRING,
          frequency: 'monthly',
          hour,
          minute: min,
          monthlyMode: 'each',
          monthDays,
          monthlyInterval,
        };
      }
    }
  }

  // Monthly "on the": M H * * 1#2 or M H * */N 1L
  if (domField === '*' && dowField !== '*' && (dowField.includes('#') || dowField.includes('L'))) {
    const min = parseInt(minField);
    const hour = parseInt(hourField);
    if (!isNaN(min) && !isNaN(hour)) {
      const monthlyInterval = parseInterval(monField) ?? 1;
      if (dowField.includes('#')) {
        const [dayStr, ordStr] = dowField.split('#');
        const weekday = parseInt(dayStr);
        const ordNum = parseInt(ordStr);
        const posMap: Record<number, OrdinalPosition> = { 1: 'first', 2: 'second', 3: 'third', 4: 'fourth' };
        if (!isNaN(weekday) && posMap[ordNum]) {
          return {
            ...DEFAULT_RECURRING,
            frequency: 'monthly',
            hour,
            minute: min,
            monthlyMode: 'onThe',
            ordinalPosition: posMap[ordNum],
            ordinalWeekday: weekday,
            monthlyInterval,
          };
        }
      } else if (dowField.endsWith('L')) {
        const weekday = parseInt(dowField.replace('L', ''));
        if (!isNaN(weekday)) {
          return {
            ...DEFAULT_RECURRING,
            frequency: 'monthly',
            hour,
            minute: min,
            monthlyMode: 'onThe',
            ordinalPosition: 'last',
            ordinalWeekday: weekday,
            monthlyInterval,
          };
        }
      }
    }
  }

  // Fallback: custom
  return { ...DEFAULT_RECURRING, frequency: 'custom', customExpression: expression };
}

function parseInterval(field: string): number | null {
  if (field === '*') return 1;
  const m = field.match(/^\*\/(\d+)$/);
  if (m) return parseInt(m[1]);
  if (field.match(/^\d+$/)) return parseInt(field);
  return null;
}

File: macos/Takt/Helpers/CronUtils.swift
import Foundation

// MARK: - Types

enum FrequencyType: String, CaseIterable {
    case interval, daily, weekly, monthly, custom

    var label: String {
        switch self {
        case .interval: return "Interval"
        case .daily: return "Daily"
        case .weekly: return "Weekly"
        case .monthly: return "Monthly"
        case .custom: return "Custom"
        }
    }
}

enum IntervalUnit: String, CaseIterable {
    case minutes, hours

    var label: String {
        switch self {
        case .minutes: return "Minutes"
        case .hours: return "Hours"
        }
    }
}

enum MonthlyMode: String {
    case each, onThe
}

enum OrdinalPosition: String, CaseIterable {
    case first, second, third, fourth, last

    var label: String { rawValue.capitalized }
}

// MARK: - RecurringState

struct RecurringState: Equatable {
    var frequency: FrequencyType = .daily
    var intervalValue: Int = 1
    var intervalUnit: IntervalUnit = .hours
    var weekdays: [Int] = [1] // 0=Sun, 1=Mon, ..., 6=Sat
    var monthlyMode: MonthlyMode = .each
    var monthDays: [Int] = [1] // 1-31
    var ordinalPosition: OrdinalPosition = .first
    var ordinalWeekday: Int = 1 // 0=Sun, ..., 6=Sat
    var hour: Int = 9
    var minute: Int = 0
    var customExpression: String = "0 * * * *"
}

let defaultRecurring = RecurringState()

// MARK: - Build Cron

func buildCron(_ state: RecurringState) -> String {
    switch state.frequency {
    case .interval:
        if state.intervalUnit == .minutes {
            return state.intervalValue == 1
                ? "* * * * *"
                : "*/\(state.intervalValue) * * * *"
        }
        // hours
        return state.intervalValue == 1
            ? "0 * * * *"
            : "0 */\(state.intervalValue) * * *"

    case .daily:
        return "\(state.minute) \(state.hour) * * *"

    case .weekly:
        let days = state.weekdays.isEmpty ? "1" : state.weekdays.sorted().map(String.init).joined(separator: ",")
        return "\(state.minute) \(state.hour) * * \(days)"

    case .monthly:
        if state.monthlyMode == .each {
            let days = state.monthDays.isEmpty ? "1" : state.monthDays.sorted().map(String.init).joined(separator: ",")
            return "\(state.minute) \(state.hour) \(days) * *"
        }
        // "On the" mode
        let ordMap: [OrdinalPosition: String] = [
            .first: "1", .second: "2", .third: "3", .fourth: "4", .last: "L"
        ]
        let ord = ordMap[state.ordinalPosition] ?? "1"
        if ord == "L" {
            return "\(state.minute) \(state.hour) * * \(state.ordinalWeekday)L"
        }
        return "\(state.minute) \(state.hour) * * \(state.ordinalWeekday)#\(ord)"

    case .custom:
        return state.customExpression
    }
}

// MARK: - Parse Cron

func parseCron(_ expression: String) -> RecurringState {
    let parts = expression.trimmingCharacters(in: .whitespaces).split(separator: " ", omittingEmptySubsequences: true).map(String.init)
    guard parts.count == 5 else {
        var s = defaultRecurring
        s.frequency = .custom
        s.customExpression = expression
        return s
    }

    let minField = parts[0]
    let hourField = parts[1]
    let domField = parts[2]
    let monField = parts[3]
    let dowField = parts[4]

    // Interval detection
    if domField == "*" && monField == "*" && dowField == "*" {
        // Minutes interval: */N * * * * or * * * * *
        if hourField == "*" && (minField == "*" || minField.range(of: #"^\*/\d+$"#, options: .regularExpression) != nil) {
            if let minInterval = parseStepInterval(minField) {
                var s = defaultRecurring
                s.frequency = .interval
                s.intervalUnit = .minutes
                s.intervalValue = minInterval
                return s
            }
        }

        // Hours interval: 0 */N * * * or 0 * * * *
        if minField == "0" && (hourField == "*" || hourField.range(of: #"^\*/\d+$"#, options: .regularExpression) != nil) {
            if let hourInterval = parseStepInterval(hourField) {
                var s = defaultRecurring
                s.frequency = .interval
                s.intervalUnit = .hours
                s.intervalValue = hourInterval
                return s
            }
        }

        // Daily: M H * * *
        if let min = Int(minField), let hour = Int(hourField) {
            var s = defaultRecurring
            s.frequency = .daily
            s.hour = hour
            s.minute = min
            return s
        }
    }

    // Weekly: M H * * 0,1,3
    if domField == "*" && dowField != "*" && !dowField.contains("#") && !dowField.contains("L") {
        if let min = Int(minField), let hour = Int(hourField), monField == "*" {
            if let weekdays = parseIntegerList(dowField, allowed: 0...6) {
                var s = defaultRecurring
                s.frequency = .weekly
                s.hour = hour
                s.minute = min
                s.weekdays = weekdays
                return s
            }
        }
    }

    // Monthly "each": M H 1,15 * *
    if domField != "*" && !domField.contains("/") && dowField == "*" && monField == "*" {
        if let min = Int(minField), let hour = Int(hourField) {
            if let monthDays = parseIntegerList(domField, allowed: 1...31) {
                var s = defaultRecurring
                s.frequency = .monthly
                s.hour = hour
                s.minute = min
                s.monthlyMode = .each
                s.monthDays = monthDays
                return s
            }
        }
    }

    // Monthly "on the": M H * * 1#2 or M H * * 1L
    if domField == "*" && dowField != "*" && (dowField.contains("#") || dowField.contains("L")) && monField == "*" {
        if let min = Int(minField), let hour = Int(hourField) {
            if dowField.contains("#") {
                let parts = dowField.split(separator: "#")
                if parts.count == 2, let weekday = Int(parts[0]), let ordNum = Int(parts[1]) {
                    let posMap: [Int: OrdinalPosition] = [1: .first, 2: .second, 3: .third, 4: .fourth]
                    if let pos = posMap[ordNum] {
                        var s = defaultRecurring
                        s.frequency = .monthly
                        s.hour = hour
                        s.minute = min
                        s.monthlyMode = .onThe
                        s.ordinalPosition = pos
                        s.ordinalWeekday = weekday
                        return s
                    }
                }
            } else if dowField.hasSuffix("L") {
                let dayStr = String(dowField.dropLast())
                if let weekday = Int(dayStr) {
                    var s = defaultRecurring
                    s.frequency = .monthly
                    s.hour = hour
                    s.minute = min
                    s.monthlyMode = .onThe
                    s.ordinalPosition = .last
                    s.ordinalWeekday = weekday
                    return s
                }
            }
        }
    }

    // Fallback: custom
    var s = defaultRecurring
    s.frequency = .custom
    s.customExpression = expression
    return s
}

// MARK: - Helpers

private func parseIntegerList(_ field: String, allowed: ClosedRange<Int>) -> [Int]? {
    var values: [Int] = []
    for part in field.split(separator: ",", omittingEmptySubsequences: false) {
        guard let value = Int(part), allowed.contains(value) else { return nil }
        values.append(value)
    }
    return values.isEmpty ? nil : values
}

private func parseStepInterval(_ field: String) -> Int? {
    if field == "*" { return 1 }
    if let range = field.range(of: #"^\*/(\d+)$"#, options: .regularExpression) {
        let match = String(field[range])
        let numStr = String(match.dropFirst(2)) // drop "*/"
        return Int(numStr)
    }
    return nil
}
